### PR TITLE
Unreal engine improvements

### PR DESCRIPTION
**Pull request by rqbacktrace**
This contains 2 improvements we need for Unreal Engine:
- Include custom attributes in native crashes (that's why we fix the BacktraceAttributes constructor)
- Include custom file attachments in managed and native crashes

### CODE REVIEW

File: backtrace-library/src/main/java/backtraceio/library/base/BacktraceBase.java [hunk 1]
```diff
@@ -371,6 +458,7 @@ public void send(BacktraceReport report, final OnServerResponseEventListener cal
         if (breadcrumbs != null) {
             breadcrumbs.processReportBreadcrumbs(report);
         }
+        processReportAttachments(report);
```
Comment (BartoszLitwiniuk) [inline backtrace-library/src/main/java/backtraceio/library/base/BacktraceBase.java:461]: maybe better name can be 'addReportAttachments'?

**Review by BartoszLitwiniuk — APPROVED** [review]
Good job 👍

File: backtrace-library/src/main/java/backtraceio/library/models/json/BacktraceAttributes.java [hunk 2]
```diff
@@ -48,9 +48,15 @@ public BacktraceAttributes(Context context, BacktraceReport report, Map<String,
             clientAttributes) {
         this.context = context;
         if (report != null) {
-            this.convertAttributes(report, clientAttributes);
+            this.convertReportAttributes(report);
             this.setExceptionAttributes(report);
         }
+        if (clientAttributes != null) {
```
Comment (konraddysput) [inline backtrace-library/src/main/java/backtraceio/library/models/json/BacktraceAttributes.java:54]: Why do we need this change in this diff? What is the purpose of this change?

File: backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java [hunk 3]
```diff
@@ -205,4 +206,11 @@ public void uncaughtException(Thread t, Throwable e) {
             Assert.fail(exception.getMessage());
         }
     }
+
+    @Test
+    public void ensureClientCustomAttributesPassedToNewBacktraceAttributes() {
+        BacktraceClient client = new BacktraceClient(context, backtraceCredentials, clientAttributes);
+        BacktraceAttributes attributes = new BacktraceAttributes(context, null, clientAttributes);
+        assertEquals(clientAttributes.get("custom-client-attribute-id"), attributes.attributes.get("custom-client-attribute-id"));
```
Comment (konraddysput) [inline backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java:214]: Why you're testing if you can get an attribute instead of test also the value of attribute? This method I believe, can return null which might be unexpected behaviour.

File: backtrace-library/src/main/java/backtraceio/library/BacktraceClient.java [hunk 4]
```diff
@@ -75,7 +131,20 @@ public BacktraceClient(Context context, BacktraceCredentials credentials,
      */
     public BacktraceClient(Context context, BacktraceCredentials credentials,
                            Database database) {
-        this(context, credentials, database, null);
+        this(context, credentials, database, (Map<String, Object>) null);
```
Comment (konraddysput) [inline backtrace-library/src/main/java/backtraceio/library/BacktraceClient.java:134]: Does it make sense to make a cast that will return null instead of passing an empty dictionary to avoid confusion? Later you need to check anyway if the dictionary exists.

File: backtrace-library/src/main/java/backtraceio/library/BacktraceDatabase.java [hunk 5]
```diff
@@ -145,7 +147,13 @@ public Boolean setupNativeIntegration(BacktraceBase client, BacktraceCredentials
         String[] values = crashpadAttributes.attributes.values().toArray(new String[0]);
 
         // Paths to Crashpad attachments
-        String[] attachmentPaths = new String[] {this.breadcrumbs.getBreadcrumbLogPath()};
+        List<String> attachmentPaths = new ArrayList<String>();
```
Comment (konraddysput) [inline backtrace-library/src/main/java/backtraceio/library/BacktraceDatabase.java:150]: Do we need to create a list? We cast it anyway to array later, so we will avoid unnecessary cast later and make code simpler with just array. We also know array size because we know how many attachments we store + additional breadcrumbs file.

File: backtrace-library/src/main/java/backtraceio/library/base/BacktraceBase.java (hunk 1 above)
Comment (konraddysput) [reply backtrace-library/src/main/java/backtraceio/library/base/BacktraceBase.java:461]: +1

File: example-app/src/main/java/backtraceio/backtraceio/MainActivity.java [hunk 6]
```diff
@@ -152,6 +170,17 @@ public void sendReport(View view) {
         backtraceClient.send(report);
     }
 
+    public void writeMyCustomFile(String filePath) {
```
Comment (konraddysput) [inline example-app/src/main/java/backtraceio/backtraceio/MainActivity.java:173]: This method should be private.

**Review by konraddysput — COMMENTED** [review]
Added few sugestions to your code to make maybe code simpler/prettier - please let me know what you think about them. I will be really happy if we can adjust few places.


Also - did you test backtraceDatabaseReport? I don't see any test with database records and new attachment system. Does backtraceDatabaseRecord store attachment path and when database try to resend a report, database report will have attachments?

Otherwise code looks fine and won't break. Thanks for this change!

File: backtrace-library/src/main/java/backtraceio/library/models/json/BacktraceAttributes.java (hunk 2 above)
Comment (rqbacktrace) [reply backtrace-library/src/main/java/backtraceio/library/models/json/BacktraceAttributes.java:54]: The purpose of this change is that we don't need to have a valid `report` to take the `clientAttributes`. If we have valid `clientAttributes`, we should include those in the new `BacktraceAttributes` even if we don't have a `report`. 

This was preventing `BacktraceAttributes` from being constructed properly for Crashpad. Because we were calling `BacktraceAttributes` constructor with a null `report`, so the `clientAttributes` were not getting passed to Crashpad. Now with this change, we do get the `clientAttributes` from Crashpad, even though we call `BacktraceAttributes` constructor with a null `report`.

File: backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java (hunk 3 above)
Comment (rqbacktrace) [reply backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java:214]: This tests that the value of `clientAttributes["custom-client-attribute-id"] == attributes.attributes["custom-client-attribute-id"]`. I will add null checks too to make it more robust.

**Review by Dautery91 — APPROVED** [review]
FYI I just took a quick skim here, pushing through so I'm not blocking you.  Looked reasonable to me, as did Konrad's feedback.

File: backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java (hunk 3 above)
Comment (konraddysput) [reply backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java:214]: but also can we be sure that `clientAttributes["custom-client-attribute-id"]` store value that it suppose to store?

File: backtrace-library/src/main/java/backtraceio/library/models/json/BacktraceAttributes.java (hunk 2 above)
Comment (konraddysput) [reply backtrace-library/src/main/java/backtraceio/library/models/json/BacktraceAttributes.java:54]: oh now I understand! thanks for explanation Rameez!

File: backtrace-library/src/main/java/backtraceio/library/BacktraceDatabase.java (hunk 5 above)
Comment (rqbacktrace) [reply backtrace-library/src/main/java/backtraceio/library/BacktraceDatabase.java:150]: If we use a Java String Array here, we get some pretty messy hard to maintain code:
`        String[] breadcrumbAttachmentPath = {this.breadcrumbs.getBreadcrumbLogPath()};
        String[] clientAttachmentPaths = (String[]) client.attachments.toArray();
        String[] attachmentPaths = new String[breadcrumbAttachmentPath.length + clientAttachmentPaths.length];
        System.arraycopy(breadcrumbAttachmentPath, 0, attachmentPaths, 0, breadcrumbAttachmentPath.length);
        System.arraycopy(clientAttachmentPaths, 0, attachmentPaths, breadcrumbAttachmentPath.length, clientAttachmentPaths.length);`
        
In general we prefer a List<String> in Java wherever possible, to avoid messy code like the above, see: https://stackoverflow.com/questions/2843366/how-to-add-new-elements-to-an-array

But we prefer a String Array for native interface, so best to work with a List<String> while we're in Java and convert to String array when we want to go to native code.

File: backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java (hunk 3 above)
Comment (rqbacktrace) [reply backtrace-library/src/androidTest/java/backtraceio/library/SettingAttributesTest.java:214]: Will do, thank you!

File: backtrace-library/src/main/java/backtraceio/library/BacktraceClient.java (hunk 4 above)
Comment (rqbacktrace) [reply backtrace-library/src/main/java/backtraceio/library/BacktraceClient.java:134]: Will do, thank you!

**Comment by rqbacktrace** [comment]
Hey thank you for the point about testing database record @konraddysput ! I actually found a bug there by making this test! Will be pushing the fix shortly

**Comment by rqbacktrace** [comment]
@BartoszLitwiniuk @konraddysput [please take a look at the latest commit in this PR](https://github.com/backtrace-labs/backtrace-android/pull/21/commits/e1fc07c45e3903380501a270308f5770e8328093). 

In this commit I fix a bug with `BacktraceData.getAttachments`. 
When `BacktraceData` is deserialized from a `BacktraceDatabaseRecord`, we lose the `Context`. But `BacktraceData.getAttachments` relies on a valid context, otherwise it breaks. So now I require `BacktraceData.getBacktraceData` to take a `Context`, so that when we deserialize the `BacktraceDatabaseRecord`, we can restore the `Context`. GSON serialization ignores the `Context`, otherwise this wouldn't be necessary anyway.

File: backtrace-library/src/main/java/backtraceio/library/base/BacktraceBase.java [hunk 7]
```diff
@@ -400,4 +487,12 @@ public void onEvent(BacktraceResult backtraceResult) {
             }
         };
     }
+
+    private void addReportAttachments(BacktraceReport report) {
```
Comment (BartoszLitwiniuk) [inline backtrace-library/src/main/java/backtraceio/library/base/BacktraceBase.java:491]: Am I correct that we will add this attachment to every report? Is this the expected behavior?

Comment (rqbacktrace) [reply backtrace-library/src/main/java/backtraceio/library/base/BacktraceBase.java:491]: Yes and yes

File: backtrace-library/src/main/java/backtraceio/library/common/BacktraceSerializeHelper.java [hunk 8]
```diff
@@ -33,6 +33,7 @@ public static String toJson(Object object) {
     }
 
     public static <T> T fromJson(String json, Class<T> type) {
-        return new Gson().fromJson(json, type);
+        Gson gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_DASHES).create();
```
Comment (BartoszLitwiniuk) [inline backtrace-library/src/main/java/backtraceio/library/common/BacktraceSerializeHelper.java:36]: maybe it is worth  to extract this line to private method in this file and use it in both places to make sure that we will not have any issues like this in the future

**Review by BartoszLitwiniuk — APPROVED** [review]
Good job!

File: backtrace-library/src/main/java/backtraceio/library/common/BacktraceSerializeHelper.java (hunk 8 above)
Comment (rqbacktrace) [reply backtrace-library/src/main/java/backtraceio/library/common/BacktraceSerializeHelper.java]: Done, thank you for the excellent advice!

File: backtrace-library/src/main/java/backtraceio/library/BacktraceDatabase.java (hunk 5 above)
Comment (rqbacktrace) [reply backtrace-library/src/main/java/backtraceio/library/BacktraceDatabase.java]: Thanks for sharing the example from Slack @konraddysput , we are now using String[] here!
